quinn-proto/src/connection/streams/recv.rs

```diff
@@ -258,40 +199,152 @@ impl Recv {
     }
 }
 
-pub(crate) type ReadResult<T> = Result<Option<DidRead<T>>, ReadError>;
-
-/// Result of a `Streams::read` call in case the stream had not ended yet
-#[derive(Debug, Eq, PartialEq, Copy, Clone)]
-#[must_use = "A frame might need to be enqueued"]
-pub(crate) struct DidRead<T> {
-    pub result: T,
-    pub max_stream_data: ShouldTransmit,
-    pub max_data: ShouldTransmit,
+/// Chunks
+pub struct Chunks<'a> {
+    id: StreamId,
+    ordered: bool,
+    streams: &'a mut Streams,
+    pending: &'a mut Retransmits,
+    state: ChunksState,
+    read: u64,
 }
 
-pub(super) type StreamReadResult<T> = Result<Option<T>, ReadError>;
+impl<'a> Chunks<'a> {
+    pub(super) fn new(
+        id: StreamId,
+        ordered: bool,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+    ) -> Result<Self, ReadableError> {
+        let entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadableError::UnknownStream),
+        };
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
-}
+        let mut recv = match entry.get().stopped {
+            true => return Err(ReadableError::UnknownStream),
+            false => entry.remove(),
+        };
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+        recv.assembler.ensure_ordering(ordered)?;
+        Ok(Self {
+            id,
+            ordered,
+            streams,
+            pending,
+            state: ChunksState::Readable(recv),
+            read: 0,
+        })
+    }
+
+    /// Next
+    ///
+    /// Should call finalize() when done calling this.
+    pub fn next(&mut self, max_length: usize) -> Result<Option<Chunk>, ReadError> {
+        let mut rs = match mem::replace(&mut self.state, ChunksState::Finalized) {
```

Comment:
Compiler says no:

```rust
error[E0507]: cannot move out of `self.state.0` which is behind a mutable reference
   --> quinn-proto/src/connection/streams/recv.rs:239:28
    |
239 |         let mut rs = match self.state {
    |                            ^^^^^^^^^^ help: consider borrowing here: `&self.state`
240 |             ChunksState::Readable(rs) => rs,
    |                                   -- data moved here
241 |             ChunksState::Error(e, st) => {
    |                                - ...and here
    |
    = note: move occurs because these variables have types that don't implement the `Copy` trait
```

Unless you know of any workarounds for this?

Reply:
Nope, just wasn't sure where rustc's line in the sand was. Thanks for having a go at it.